velox/exec/tests/HashPartitionFunctionTest.cpp

```diff
@@ -114,3 +114,36 @@ TEST_F(HashPartitionFunctionTest, HashPartitionFunction) {
     EXPECT_EQ(4, functionWithBits2.numPartitions());
   }
 }
+
+TEST_F(HashPartitionFunctionTest, spec) {
+  Type::registerSerDe();
+  auto& registry = DeserializationWithContextRegistryForSharedPtr();
+  registry.Register(
+      "ConstantTypedExpr", facebook::velox::core::ConstantTypedExpr::create);
+  
+  RowTypePtr outputType(
+      ROW({"c0", "c1", "c2", "c3", "c4"},
+          {BIGINT(), SMALLINT(), INTEGER(), BIGINT(), VARCHAR()}));
+  auto hashSpec = std::make_unique<exec::HashPartitionFunctionSpec>(
+      outputType,
+      std::vector<column_index_t>{0, 1, 2, 3, 4},
+      std::vector<VectorPtr>{makeConstant(1, 1)});
+
+  auto serialized = hashSpec->serialize();
+  auto copy = HashPartitionFunctionSpec::deserialize(serialized, pool());
+  ASSERT_EQ(serialized["constants"].size(), 1);
+  ASSERT_EQ(hashSpec->toString(), copy->toString());
+}
+
+TEST_F(HashPartitionFunctionTest, specConstValues) {
+  Type::registerSerDe();
+  RowTypePtr outputType(ROW({"c0", "c1"}, {BIGINT(), SMALLINT()}));
+  auto hashSpec = std::make_unique<exec::HashPartitionFunctionSpec>(
+      outputType,
+      std::vector<column_index_t>{0, 1},
+      std::vector<VectorPtr>{nullptr});
```

Comment:
Can this be null? Shall we test std::vector<VectorPtr> is empty instead?

Reply:
I imagine that there can be 5 keys, some of which are constant and some not. For example, first 2 keys are constant or last 2 keys are constant or every other key is constant or none of the keys are constant. Let's make sure to test all these cases.